src/pocketmine/form/Form.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+/**
+ * API for Minecraft: Bedrock custom UI (forms)
+ */
+namespace pocketmine\form;
+
+use pocketmine\Player;
+
+/**
+ * Base class for a custom form. Forms are serialized to JSON data to be sent to clients.
+ */
+abstract class Form implements \JsonSerializable{
+
+	const TYPE_MODAL = "modal";
+	const TYPE_MENU = "form";
+	const TYPE_CUSTOM_FORM = "custom_form";
+
+	/** @var string */
+	protected $title = "";
+
+	public function __construct(string $title){
+		$this->title = $title;
+	}
+
+	/**
+	 * Returns the type used to show this form to clients
+	 * @return string
+	 */
+	abstract public function getType() : string;
+
+	/**
+	 * Returns the text shown on the form title-bar.
+	 * @return string
+	 */
+	public function getTitle() : string{
+		return $this->title;
+	}
+
+	/**
+	 * Handles a form response from a player. Plugins should not override this method, override {@link onSubmit}
+	 * instead.
+	 *
+	 * @param Player $player
+	 * @param mixed  $data
+	 */
+	abstract public function handleResponse(Player $player, $data) : void;
+
+	/**
+	 * Called when a player submits this form. Each form type usually has its own methods for getting relevant data from
+	 * them.
+	 *
+	 * Plugins should extend the class and override this function and add their own code to handle form responses as
+	 * they wish.
+	 *
+	 * @param Player $player
+	 */
+	abstract public function onSubmit(Player $player) : void;
+
+	/**
+	 * Clears response data from a form, useful if you want to reuse the same form object several times.
+	 */
+	public function clearResponseData() : void{
```

Comment:
This method is very confusing. It is not called by the server, but relies on the developer to call it.

This leads to inconsistency:
1. Is a Form object expected to be reused for a single player?
2. Is a Form object expected to be reused for multiple players?
3. Can the developer use the form values outside onSubmit()?

The answer is:
* If the same Form instance is only sent once, you can use the form values outside onSubmit() and hold the Form instance forever.
* If the same Form instance is sent more than once (same player or not), it would be a very bad practice to depend on the value held in the Form instance outside onSubmit(), even in the same tick.

It is particularly tempting to hold a CustomForm instance and use the values there later, since the data cannot be easily extracted. I can foresee this is going to cause a lot of concurrency bugs that can be very difficult to identify. Therefore, it is suggested that, starting from the core, one of either approach ("resend same instance" or "new instance each resend") be adopted to avoid confusion.

* Adopting the "resend same instance" approach should automatically call the `clearResponseData()` method after each onSubmit() call.
* Adopting the "new instance each resend" approach should remove this method and add a flag to identify a Form as sent and throw an exception when attempting to use sendForm() with it.

Deciding which approach to adopt should require public consultation.